#### Objective of this project
To build a model to accurately classify a piece of news as REAL or FAKE. Using sklearn, we build a TfidfVectorizer on our dataset. Then, we initialize a PassiveAggressive Classifier and fit the model. In the end, the accuracy score and the confusion matrix tell us how well our model fares.

#### Dataset
The dataset we’ll use for this python project has a shape of 7796×4. The first column identifies the news, the second and third are the title and text, and the fourth column has labels denoting whether the news is REAL or FAKE. 

#### What is Fake News?
A type of yellow journalism, fake news encapsulates pieces of news that may be hoaxes and is generally spread through social media and other online media. This is often done to further or impose certain ideas and is often achieved with political agendas. Such news items may contain false and/or exaggerated claims, and may end up being viralized by algorithms, and users may end up in a filter bubble.

#### What is a PassiveAggressiveClassifier?
Passive Aggressive algorithms are online learning algorithms. Such an algorithm remains passive for a correct classification outcome, and turns aggressive in the event of a miscalculation, updating and adjusting. Unlike most other algorithms, it does not converge. Its purpose is to make updates that correct the loss, causing very little change in the norm of the weight vector.

#### What is a TF-IDF Vectorizer?
- TF (Term Frequency): The number of times a word appears in a document is its Term Frequency. A higher value means a term appears more often than others, and so, the document is a good match when the term is part of the search terms.

- IDF (Inverse Document Frequency): Words that occur many times a document, but also occur many times in many others, may be irrelevant. IDF is a measure of how significant a term is in the entire corpus.

In [1]:
import numpy as np
import pandas as pd
import itertools
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

In [2]:
# Read the dataset
df = pd.read_csv('dataset/fake_news_detection/news.csv')

In [3]:
df.sample(5)

Unnamed: 0                                              title  \
3062       10149  Texas County Enacts "Emergency Paper Ballots" ...   
4828         105  Let’s make the South stop lying: The right’s w...   
1239        1406    GOP insiders: It was a rough night for Ted Cruz   
2003         685  Bernie Sanders, Hillary Clinton battle for Cal...   
1668         224     Ryan flexes muscles in Ways and Means showdown   

                                                   text label  
3062    Texas County Enacts "Emergency Paper Ballots...  FAKE  
4828  The final vote in the state House was a lopsid...  REAL  
1239  Killing Obama administration rules, dismantlin...  REAL  
2003  San Diego (CNN) Hillary Clinton is back in the...  REAL  
1668  Top Dems want White House to call off Part B d...  REAL

In [4]:
# Split the dataset into training and test sets.
x_train,x_test,y_train,y_test=train_test_split(df['text'], df['label'], test_size=0.2, random_state=42)

Stop words are the most common words in a language that are to be filtered out before processing the natural language data. 

TfidfVectorizer turns a collection of raw documents into a matrix of TF-IDF features.

Let’s initialize a TfidfVectorizer with stop words from the English language and a maximum document frequency of 0.7 (terms with a higher document frequency will be discarded). 

In [5]:
tfidf_vectorizer = TfidfVectorizer(stop_words='english', max_df=0.7)

# Fit and transform the train set and then use the learned tfidf model to transform test set
tfidf_train=tfidf_vectorizer.fit_transform(x_train) 
tfidf_test=tfidf_vectorizer.transform(x_test)

Now initialize a PassiveAggressiveClassifier. We’ll fit this on tfidf_train and y_train.

Then, we’ll predict on the test set from the TfidfVectorizer and calculate the accuracy with accuracy_score() from sklearn.metrics.

In [6]:
# Train the model on train set
pac = PassiveAggressiveClassifier(max_iter=100)
pac.fit(tfidf_train, y_train)

PassiveAggressiveClassifier(max_iter=100)

In [7]:
# Predict on the test set and calculate accuracy
y_pred=pac.predict(tfidf_test)
score=accuracy_score(y_test, y_pred)
print(f'Accuracy: {round(score*100,2)}%')

Accuracy: 93.53%


Let’s print out a confusion matrix to gain insight into the number of false and true negatives and positives.

In [8]:
confusion_matrix(y_test, y_pred, labels=['FAKE','REAL'])

array([[588,  40],
       [ 42, 597]], dtype=int64)

#### Summary

We took a political news dataset, implemented a TfidfVectorizer, initialized a PassiveAggressiveClassifier, and fit our model. We ended up obtaining an accuracy of 93.5% in magnitude.